### PR TITLE
Eyes automation: Spin and Beat jump

### PR DESCRIPTION
**Pull request by synox**
I created two GLSL programs that automatically control the eyes. The colors are kind of random, to be tested. 

We should also test if the Moving Heads are fast enough to follow the motion and color. 

Demo: 
demo

### CODE REVIEW

**Comment by xian** [comment]
Love it @synox! I'll look closer asap.

Theoretically the simulator is [modeling max hardware travel time](https://github.com/baaahs/sparklemotion/blob/d1a740518d0ce812c5d0369affb42ff616760257/src/commonMain/kotlin/baaahs/visualizer/movers/State.kt#L15) based on mfr specs but... yeah we should test it out. ;-)

File: src/commonMain/resources/templates/shows/copyForReview-spin.glsl [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    return fract(sin(dot(co, vec2(12.9898, 78.233))) * 43758.5453);
+}
+
+#define PI 3.14159265358979323846
+
+
+// @param params moving-head-params
+void main(out MovingHeadParams params) {
+    bool isLeft = fixtureInfo.position.z < 0.;
+    float rotation = 0.5 + (cos(speed*time/3.)+sin(speed*time/2.)) / 2.0 / 2.0;// [0...1]
```
Comment (synox) [inline src/commonMain/resources/templates/shows/copyForReview-spin.glsl:32]: turns end-to-end with some wiggle in between. see https://graphtoy.com/?f1(x,t)=0.5%20+%20(cos(10*x/3.)+sin(10*x/2.))%20/%202.0%20/%202.0&v1=true&f2(x,t)=&v2=true&f3(x,t)=&v3=true&f4(x,t)=&v4=true&f5(x,t)=&v5=false&f6(x,t)=&v6=true&grid=1&coords=2.8945805231935795,0.2949890979050782,6.157897418768507

File: src/commonMain/resources/templates/shows/copyForReview-spin.glsl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    //rotation = pos;
```
Comment (synox) [inline src/commonMain/resources/templates/shows/copyForReview-spin.glsl:33]: TODO: cleanup and add comments

File: src/commonMain/resources/templates/shows/copyForReview-manual.glsl [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    float colorWheel;
+    float dimmer;
+};
+
+#define PI 3.14159265358979323846
+#define DISTANCE_FORWARD 500.0 // Inches
+#define DISTANCE_SIDEWAYS 300.0 // Inches to each side
+
+uniform FixtureInfo fixtureInfo;
+uniform vec2 floorPosition; // @@XyPad
```
Comment (xian) [inline src/commonMain/resources/templates/shows/copyForReview-manual.glsl:19]: Nice! Floor works great for headlights mode, where eyes should point at the road ahead.

Thinking about how to aim in other cases, like at the disco balls, which are just a bit below and aft of eye level and off to one side. Maybe consider a few modes for the position, like floor, sky, and something like horizon, which would be a cylinder surrounding us?

File: src/commonMain/resources/templates/shows/copyForReview-manual.glsl [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+struct MovingHeadParams {
+    float pan;
+    float tilt;
+    float colorWheel;
+    float dimmer;
+};
+
+#define PI 3.14159265358979323846
+#define DISTANCE_FORWARD 500.0 // Inches
```
Comment (xian) [inline src/commonMain/resources/templates/shows/copyForReview-manual.glsl:15]: You can get model bounds with something like this:
```glsl
struct ModelInfo {
    vec3 center;            
    vec3 extents;
}
uniform ModelInfo modelInfo;

// bottom-left-far == modelInfo.center - modelInfo.extents / 2;
// top-right-near == modelInfo.center + modelInfo.extents / 2;
```

File: src/commonMain/resources/templates/shows/copyForReview-beat-jump.glsl [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        float randomForward = rand(vec2(1, float(getTimeOfLastBeatSet())));
+        float randomSideways = rand(vec2(2, float(getTimeOfLastBeatSet())));
+
+        vec3 target = vec3(randomForward  * DISTANCE_FORWARD,
+        0.0,
+        mix(-DISTANCE_SIDEWAYS, DISTANCE_SIDEWAYS, randomSideways)
+        );
+
+        // correct for the yaw-rotation of the fixture
+        vec3 rotatedTarget = rotateAroundYAxis(target, fixtureInfo.rotation.y);
```
Comment (xian) [inline src/commonMain/resources/templates/shows/copyForReview-beat-jump.glsl:77]: There's some matrix magic you could do here instead with `fixtureInfo.transformation` but I'm fuzzy on exactly what. Something like, ish:

```glsl
vec3 targetInFixtureSpace = (inverse(fixtureInfo.transformation) * vec4(target, 0.)).xyz;
```

**Comment by synox** [comment]
@xian if you approve this, I'll merge it in

**Comment by synox** [comment]
replaced by https://github.com/baaahs/sparklemotion/pull/546

**Pull request closed without merging**